File: android/app/src/main/java/my/ssdid/wallet/feature/credentials/CredentialOfferScreen.kt
package my.ssdid.wallet.feature.credentials

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Schedule
import my.ssdid.wallet.domain.model.Identity
import my.ssdid.wallet.ui.components.HapticManager
import my.ssdid.wallet.ui.theme.*

@Composable
fun CredentialOfferScreen(
    onBack: () -> Unit,
    onComplete: () -> Unit,
    viewModel: CredentialOfferViewModel = hiltViewModel()
) {
    val state by viewModel.state.collectAsState()
    val view = LocalView.current

    LaunchedEffect(state) {
        if (state is CredentialOfferUiState.Success) HapticManager.success(view)
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPrimary)
            .statusBarsPadding().navigationBarsPadding()
    ) {
        // Header
        Row(
            Modifier.padding(start = 8.dp, end = 20.dp, top = 12.dp, bottom = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(onClick = onBack) {
                Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back", tint = TextPrimary)
            }
            Spacer(Modifier.width(4.dp))
            Text("Credential Offer", style = MaterialTheme.typography.titleLarge)
        }

        when (val currentState = state) {
            is CredentialOfferUiState.Loading -> LoadingContent("Processing offer...")

            is CredentialOfferUiState.ReviewingOffer -> ReviewingOfferContent(
                issuerName = currentState.issuerName,
                credentialTypes = currentState.credentialTypes,
                identities = currentState.identities,
                selectedIdentity = currentState.selectedIdentity,
                onSelectIdentity = viewModel::selectIdentity,
                onAccept = viewModel::acceptOffer,
                onDecline = { viewModel.decline(); onBack() }
            )

            is CredentialOfferUiState.PinEntry -> PinEntryContent(
                description = currentState.description,
                length = currentState.length,
                inputMode = currentState.inputMode,
                onSubmit = viewModel::submitPin,
                onCancel = { viewModel.decline(); onBack() }
            )

            is CredentialOfferUiState.Processing -> LoadingContent("Requesting credential...")

            is CredentialOfferUiState.Success -> SuccessContent(onComplete = onComplete)

            is CredentialOfferUiState.Deferred -> DeferredContent(
                transactionId = currentState.transactionId,
                onDone = onComplete
            )

            is CredentialOfferUiState.Error -> ErrorContent(
                message = currentState.message,
                onBack = onBack
            )
        }
    }
}

@Composable
private fun ColumnScope.LoadingContent(message: String) {
    Spacer(Modifier.weight(1f))
    Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            CircularProgressIndicator(color = Accent)
            Spacer(Modifier.height(16.dp))
            Text(message, color = TextSecondary, fontSize = 14.sp)
        }
    }
    Spacer(Modifier.weight(1f))
}

@Composable
private fun ColumnScope.ReviewingOfferContent(
    issuerName: String,
    credentialTypes: List<String>,
    identities: List<Identity>,
    selectedIdentity: Identity?,
    onSelectIdentity: (Identity) -> Unit,
    onAccept: () -> Unit,
    onDecline: () -> Unit
) {
    // Issuer info card
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = BgCard)
    ) {
        Column(Modifier.padding(14.dp)) {
            Text("ISSUER", style = MaterialTheme.typography.labelMedium)
            Spacer(Modifier.height(8.dp))
            Text(issuerName, fontSize = 13.sp, color = TextPrimary, fontFamily = FontFamily.Monospace, maxLines = 2)
        }
    }

    Spacer(Modifier.height(12.dp))

    // Credential types
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = BgCard)
    ) {
        Column(Modifier.padding(14.dp)) {
            Text("CREDENTIALS OFFERED", style = MaterialTheme.typography.labelMedium)
            Spacer(Modifier.height(8.dp))
            credentialTypes.forEach { type ->
                Text(type, fontSize = 13.sp, color = Accent, modifier = Modifier.padding(vertical = 2.dp))
            }
        }
    }

    Spacer(Modifier.height(12.dp))

    // Identity selector
    Text(
        "SELECT IDENTITY",
        style = MaterialTheme.typography.labelMedium,
        modifier = Modifier.padding(horizontal = 20.dp)
    )
    Spacer(Modifier.height(8.dp))

    LazyColumn(
        modifier = Modifier
            .weight(1f)
            .padding(horizontal = 20.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        items(identities) { identity ->
            val isSelected = selectedIdentity?.keyId == identity.keyId
            Card(
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable { onSelectIdentity(identity) },
                shape = RoundedCornerShape(12.dp),
                colors = CardDefaults.cardColors(
                    containerColor = if (isSelected) AccentDim else BgCard
                )
            ) {
                Row(Modifier.padding(14.dp), verticalAlignment = Alignment.CenterVertically) {
                    RadioButton(
                        selected = isSelected,
                        onClick = { onSelectIdentity(identity) },
                        colors = RadioButtonDefaults.colors(selectedColor = Accent)
                    )
                    Spacer(Modifier.width(8.dp))
                    Column {
                        Text(identity.name, style = MaterialTheme.typography.titleMedium)
                        Text(
                            identity.did,
                            fontSize = 11.sp,
                            color = TextTertiary,
                            fontFamily = FontFamily.Monospace,
                            maxLines = 1
                        )
                    }
                }
            }
        }

        if (identities.isEmpty()) {
            item {
                Box(
                    Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(12.dp))
                        .background(BgCard)
                        .padding(24.dp),
                    contentAlignment = Alignment.Center
                ) {
                    Text("No identities available", color = TextSecondary)
                }
            }
        }
    }

    // Accept/Decline buttons
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(20.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        OutlinedButton(
            onClick = onDecline,
            modifier = Modifier.weight(1f),
            shape = RoundedCornerShape(12.dp),
            colors = ButtonDefaults.outlinedButtonColors(contentColor = Danger)
        ) {
            Text("Decline", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
        }
        Button(
            onClick = onAccept,
            modifier = Modifier.weight(1f),
            enabled = selectedIdentity != null,
            shape = RoundedCornerShape(12.dp),
            colors = ButtonDefaults.buttonColors(containerColor = Accent)
        ) {
            Text("Accept", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
        }
    }
}

@Composable
private fun ColumnScope.PinEntryContent(
    description: String?,
    length: Int,
    inputMode: String,
    onSubmit: (String) -> Unit,
    onCancel: () -> Unit
) {
    var pinValue by remember { mutableStateOf("") }

    Spacer(Modifier.weight(1f))
    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 32.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text("Transaction Code Required", style = MaterialTheme.typography.headlineSmall, color = TextPrimary)
        Spacer(Modifier.height(8.dp))
        Text(
            description ?: "Please enter the transaction code provided by the issuer.",
            color = TextSecondary,
            fontSize = 14.sp
        )
        Spacer(Modifier.height(24.dp))
        OutlinedTextField(
            value = pinValue,
            onValueChange = { newValue ->
                if (length <= 0 || newValue.length <= length) {
                    pinValue = newValue
                }
            },
            label = { Text("Transaction Code") },
            keyboardOptions = KeyboardOptions(
                keyboardType = if (inputMode == "numeric") KeyboardType.Number else KeyboardType.Text
            ),
            singleLine = true,
            modifier = Modifier.fillMaxWidth()
        )
        if (length > 0) {
            Spacer(Modifier.height(4.dp))
            Text("${pinValue.length}/$length characters", fontSize = 12.sp, color = TextTertiary)
        }
    }
    Spacer(Modifier.weight(1f))

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(20.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        OutlinedButton(
            onClick = onCancel,
            modifier = Modifier.weight(1f),
            shape = RoundedCornerShape(12.dp),
            colors = ButtonDefaults.outlinedButtonColors(contentColor = Danger)
        ) {
            Text("Cancel", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
        }
        Button(
            onClick = { onSubmit(pinValue) },
            modifier = Modifier.weight(1f),
            enabled = pinValue.isNotEmpty() && (length <= 0 || pinValue.length == length),
            shape = RoundedCornerShape(12.dp),
            colors = ButtonDefaults.buttonColors(containerColor = Accent)
        ) {
            Text("Submit", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
        }
    }
}

@Composable
private fun ColumnScope.SuccessContent(onComplete: () -> Unit) {
    Spacer(Modifier.weight(1f))
    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 32.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(
            Modifier
                .size(72.dp)
                .clip(RoundedCornerShape(20.dp))
                .background(SuccessDim),
            contentAlignment = Alignment.Center
        ) {
            Icon(Icons.Default.Check, contentDescription = "Success", modifier = Modifier.size(32.dp), tint = Success)
        }
        Spacer(Modifier.height(20.dp))
        Text("Credential Received", style = MaterialTheme.typography.headlineSmall, color = TextPrimary)
        Spacer(Modifier.height(8.dp))
        Text("The credential has been stored in your wallet.", color = TextSecondary, fontSize = 14.sp)
    }
    Spacer(Modifier.weight(1f))
    Button(
        onClick = onComplete,
        modifier = Modifier
            .fillMaxWidth()
            .padding(20.dp),
        shape = RoundedCornerShape(12.dp),
        colors = ButtonDefaults.buttonColors(containerColor = Success)
    ) {
        Text("Done", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
    }
}

@Composable
private fun ColumnScope.DeferredContent(transactionId: String, onDone: () -> Unit) {
    Spacer(Modifier.weight(1f))
    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 32.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(
            Modifier
                .size(72.dp)
                .clip(RoundedCornerShape(20.dp))
                .background(WarningDim),
            contentAlignment = Alignment.Center
        ) {
            Icon(Icons.Default.Schedule, contentDescription = "Pending", modifier = Modifier.size(32.dp), tint = Warning)
        }
        Spacer(Modifier.height(20.dp))
        Text("Credential Pending", style = MaterialTheme.typography.headlineSmall, color = TextPrimary)
        Spacer(Modifier.height(8.dp))
        Text(
            "Your credential is being processed and will be available later.",
            color = TextSecondary,
            fontSize = 14.sp
        )
        Spacer(Modifier.height(12.dp))
        Text(
            "Transaction ID: $transactionId",
            fontSize = 11.sp,
            color = TextTertiary,
            fontFamily = FontFamily.Monospace
        )
    }
    Spacer(Modifier.weight(1f))
    Button(
        onClick = onDone,
        modifier = Modifier
            .fillMaxWidth()
            .padding(20.dp),
        shape = RoundedCornerShape(12.dp),
        colors = ButtonDefaults.buttonColors(containerColor = Warning)
    ) {
        Text("Done", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
    }
}

@Composable
private fun ColumnScope.ErrorContent(message: String, onBack: () -> Unit) {
    Spacer(Modifier.weight(1f))
    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 32.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(
            Modifier
                .size(72.dp)
                .clip(RoundedCornerShape(20.dp))
                .background(DangerDim),
            contentAlignment = Alignment.Center
        ) {
            Icon(Icons.Default.Close, contentDescription = "Error", modifier = Modifier.size(32.dp), tint = Danger)
        }
        Spacer(Modifier.height(20.dp))
        Text("Offer Failed", style = MaterialTheme.typography.headlineSmall, color = TextPrimary)
        Spacer(Modifier.height(8.dp))
        Text(message, color = TextSecondary, fontSize = 14.sp)
    }
    Spacer(Modifier.weight(1f))
    Button(
        onClick = onBack,
        modifier = Modifier
            .fillMaxWidth()
            .padding(20.dp),
        shape = RoundedCornerShape(12.dp),
        colors = ButtonDefaults.buttonColors(containerColor = Danger)
    ) {
        Text("Go Back", fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
    }
}

File: ios/SsdidWallet/Feature/Credentials/CredentialOfferScreen.swift
import SwiftUI

struct CredentialOfferScreen: View {
    @Environment(AppRouter.self) private var router
    @EnvironmentObject private var services: ServiceContainer

    let issuerUrl: String
    let offerId: String

    @State private var viewModel: CredentialOfferViewModel?
    @State private var pinText = ""

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack(spacing: 4) {
                Button { router.pop() } label: {
                    Image(systemName: "chevron.left")
                        .foregroundStyle(Color.textPrimary)
                        .font(.system(size: 20))
                }
                .padding(.leading, 8)

                Text("Credential Offer")
                    .font(.ssdidHeadline)
                    .foregroundStyle(Color.textPrimary)

                Spacer()
            }
            .padding(.vertical, 12)
            .padding(.trailing, 20)

            if let viewModel {
                stateContent(viewModel)
            } else {
                loadingView(message: "Initializing...")
            }
        }
        .background(Color.bgPrimary)
        .onAppear {
            if viewModel == nil {
                let metadataResolver = IssuerMetadataResolver()
                let tokenClient = TokenClient()
                let nonceManager = NonceManager()
                let transport = OpenId4VciTransport()
                let handler = OpenId4VciHandler(
                    metadataResolver: metadataResolver,
                    tokenClient: tokenClient,
                    nonceManager: nonceManager,
                    transport: transport,
                    vcStorage: services.storage
                )
                let vm = CredentialOfferViewModel(handler: handler, vault: services.vault)
                viewModel = vm

                if !issuerUrl.isEmpty {
                    let offerUri = buildOfferUri(issuerUrl: issuerUrl, offerId: offerId)
                    vm.processOffer(rawUri: offerUri)
                } else if !offerId.isEmpty {
                    // offerId may contain the full offer data from a deep link
                    vm.processOfferJson(json: offerId)
                }
            }
        }
    }

    // MARK: - State Content

    @ViewBuilder
    private func stateContent(_ vm: CredentialOfferViewModel) -> some View {
        switch vm.state {
        case .loading:
            loadingView(message: "Fetching offer details...")

        case .reviewingOffer(let data):
            reviewContent(data: data, vm: vm)

        case .pinEntry(let pinData):
            pinEntryContent(pinData: pinData, vm: vm)

        case .processing:
            loadingView(message: "Accepting credential...")

        case .success:
            successView()

        case .deferred(let transactionId):
            deferredView(transactionId: transactionId)

        case .error(let message):
            errorView(message: message)
        }
    }

    // MARK: - Loading

    @ViewBuilder
    private func loadingView(message: String) -> some View {
        Spacer()
        VStack(spacing: 16) {
            ProgressView()
                .progressViewStyle(CircularProgressViewStyle(tint: Color.ssdidAccent))
            Text(message)
                .font(.system(size: 14))
                .foregroundStyle(Color.textSecondary)
        }
        Spacer()
    }

    // MARK: - Reviewing Offer

    @ViewBuilder
    private func reviewContent(data: CredentialOfferViewModel.ReviewData, vm: CredentialOfferViewModel) -> some View {
        ScrollView {
            LazyVStack(alignment: .leading, spacing: 12) {
                // Offer details card
                VStack(alignment: .leading, spacing: 0) {
                    Text("OFFER DETAILS")
                        .font(.ssdidCaption)
                        .foregroundStyle(Color.textTertiary)
                    Spacer().frame(height: 8)
                    Text("Issuer")
                        .font(.system(size: 11))
                        .foregroundStyle(Color.textTertiary)
                    Text(data.issuerName)
                        .font(.ssdidMono)
                        .foregroundStyle(Color.textPrimary)
                        .lineLimit(1)
                    Spacer().frame(height: 6)
                    Text("Credential Type")
                        .font(.system(size: 11))
                        .foregroundStyle(Color.textTertiary)

                    ForEach(data.credentialTypes, id: \.self) { type in
                        let isSelected = type == data.selectedConfigId
                        Button {
                            vm.selectConfigId(type)
                        } label: {
                            HStack(spacing: 6) {
                                if data.credentialTypes.count > 1 {
                                    Image(systemName: isSelected ? "largecircle.fill.circle" : "circle")
                                        .foregroundStyle(isSelected ? Color.ssdidAccent : Color.textTertiary)
                                        .font(.system(size: 14))
                                }
                                Text(type)
                                    .font(.system(size: 13))
                                    .foregroundStyle(isSelected ? Color.ssdidAccent : Color.textPrimary)
                            }
                        }
                    }

                    if data.requiresPin {
                        Spacer().frame(height: 6)
                        HStack(spacing: 4) {
                            Image(systemName: "lock.fill")
                                .font(.system(size: 10))
                                .foregroundStyle(Color.warning)
                            Text("PIN required")
                                .font(.system(size: 11))
                                .foregroundStyle(Color.warning)
                        }
                    }
                }
                .ssdidCard()

                // Identity selector
                Text("SELECT IDENTITY")
                    .font(.ssdidCaption)
                    .foregroundStyle(Color.textTertiary)

                ForEach(data.identities) { identity in
                    let isSelected = data.selectedIdentity?.keyId == identity.keyId
                    Button { vm.selectIdentity(identity) } label: {
                        HStack(spacing: 8) {
                            Image(systemName: isSelected ? "largecircle.fill.circle" : "circle")
                                .foregroundStyle(isSelected ? Color.ssdidAccent : Color.textTertiary)
                            VStack(alignment: .leading, spacing: 2) {
                                Text(identity.name)
                                    .font(.system(size: 14, weight: .medium))
                                    .foregroundStyle(Color.textPrimary)
                                Text(identity.did)
                                    .font(.system(size: 11, design: .monospaced))
                                    .foregroundStyle(Color.textTertiary)
                                    .lineLimit(1)
                            }
                            Spacer()
                        }
                        .padding(14)
                        .background(isSelected ? Color.accentDim : Color.bgCard)
                        .cornerRadius(12)
                    }
                }

                if data.identities.isEmpty {
                    Text("No identities available")
                        .font(.ssdidBody)
                        .foregroundStyle(Color.textSecondary)
                        .frame(maxWidth: .infinity)
                        .padding(24)
                        .background(Color.bgCard)
                        .cornerRadius(12)
                }
            }
            .padding(.horizontal, 20)
        }

        // Accept/Reject buttons
        HStack(spacing: 12) {
            Button { vm.decline() } label: { Text("Reject") }
                .buttonStyle(.ssdidDanger)

            Button { vm.acceptOffer() } label: { Text("Accept") }
                .buttonStyle(.ssdidPrimary(enabled: data.selectedIdentity != nil))
                .disabled(data.selectedIdentity == nil)
        }
        .padding(.horizontal, 20)
        .padding(.vertical, 12)
    }

    // MARK: - PIN Entry

    @ViewBuilder
    private func pinEntryContent(pinData: CredentialOfferViewModel.PinData, vm: CredentialOfferViewModel) -> some View {
        Spacer()
        VStack(spacing: 20) {
            Image(systemName: "lock.shield")
                .font(.system(size: 48))
                .foregroundStyle(Color.ssdidAccent)

            Text("Enter PIN")
                .font(.ssdidHeadline)
                .foregroundStyle(Color.textPrimary)

            if let description = pinData.description {
                Text(description)
                    .font(.system(size: 14))
                    .foregroundStyle(Color.textSecondary)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 32)
            }

            TextField("PIN", text: $pinText)
                .textFieldStyle(.roundedBorder)
                .keyboardType(pinData.inputMode == "numeric" ? .numberPad : .default)
                .multilineTextAlignment(.center)
                .frame(maxWidth: 200)
                .font(.system(size: 24, weight: .medium, design: .monospaced))

            if pinData.length > 0 {
                Text("\(pinData.length) characters required")
                    .font(.system(size: 12))
                    .foregroundStyle(Color.textTertiary)
            }
        }
        Spacer()

        HStack(spacing: 12) {
            Button {
                pinText = ""
                // Go back to reviewing
                vm.processOffer(rawUri: buildOfferUri(issuerUrl: issuerUrl, offerId: offerId))
            } label: { Text("Cancel") }
                .buttonStyle(.ssdidDanger)

            Button { vm.submitPin(pinText) } label: { Text("Submit") }
                .buttonStyle(.ssdidPrimary(enabled: !pinText.isEmpty))
                .disabled(pinText.isEmpty)
        }
        .padding(.horizontal, 20)
        .padding(.vertical, 12)
    }

    // MARK: - Success

    @ViewBuilder
    private func successView() -> some View {
        Spacer()
        VStack(spacing: 0) {
            ZStack {
                RoundedRectangle(cornerRadius: 20)
                    .fill(Color.successDim)
                    .frame(width: 72, height: 72)
                Image(systemName: "checkmark")
                    .font(.system(size: 32, weight: .bold))
                    .foregroundStyle(Color.success)
            }
            Spacer().frame(height: 20)
            Text("Credential Accepted")
                .font(.ssdidHeadline)
                .foregroundStyle(Color.textPrimary)
            Spacer().frame(height: 8)
            Text("The credential has been stored in your wallet.")
                .font(.system(size: 14))
                .foregroundStyle(Color.textSecondary)
                .multilineTextAlignment(.center)
        }
        .padding(.horizontal, 32)
        Spacer()
        Button {
            HapticManager.notification(.success)
            router.pop()
        } label: { Text("Done") }
            .buttonStyle(.ssdidPrimary)
            .padding(20)
    }

    // MARK: - Deferred

    @ViewBuilder
    private func deferredView(transactionId: String) -> some View {
        Spacer()
        VStack(spacing: 0) {
            ZStack {
                RoundedRectangle(cornerRadius: 20)
                    .fill(Color.accentDim)
                    .frame(width: 72, height: 72)
                Image(systemName: "clock")
                    .font(.system(size: 32, weight: .bold))
                    .foregroundStyle(Color.ssdidAccent)
            }
            Spacer().frame(height: 20)
            Text("Credential Pending")
                .font(.ssdidHeadline)
                .foregroundStyle(Color.textPrimary)
            Spacer().frame(height: 8)
            Text("Your credential request has been submitted. You will be notified when it is ready.")
                .font(.system(size: 14))
                .foregroundStyle(Color.textSecondary)
                .multilineTextAlignment(.center)
            Spacer().frame(height: 12)
            Text("Transaction: \(transactionId)")
                .font(.system(size: 11, design: .monospaced))
                .foregroundStyle(Color.textTertiary)
                .lineLimit(1)
        }
        .padding(.horizontal, 32)
        Spacer()
        Button { router.pop() } label: { Text("Done") }
            .buttonStyle(.ssdidPrimary)
            .padding(20)
    }

    // MARK: - Error

    @ViewBuilder
    private func errorView(message: String) -> some View {
        Spacer()
        VStack(spacing: 0) {
            ZStack {
                RoundedRectangle(cornerRadius: 20)
                    .fill(Color.dangerDim)
                    .frame(width: 72, height: 72)
                Image(systemName: "xmark")
                    .font(.system(size: 32, weight: .bold))
                    .foregroundStyle(Color.danger)
            }
            Spacer().frame(height: 20)
            Text("Offer Failed")
                .font(.ssdidHeadline)
                .foregroundStyle(Color.textPrimary)
            Spacer().frame(height: 8)
            Text(message)
                .font(.system(size: 14))
                .foregroundStyle(Color.textSecondary)
                .multilineTextAlignment(.center)
        }
        .padding(.horizontal, 32)
        Spacer()
        Button { router.pop() } label: { Text("Go Back") }
            .buttonStyle(.ssdidDanger)
            .padding(20)
    }

    // MARK: - Helpers

    private func buildOfferUri(issuerUrl: String, offerId: String) -> String {
        let offerJson = """
        {"credential_issuer":"\(issuerUrl)","credential_configuration_ids":["\(offerId)"],"grants":{"urn:ietf:params:oauth:grant-type:pre-authorized_code":{"pre-authorized_code":"\(offerId)"}}}
        """
        guard let encoded = offerJson.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
            return "openid-credential-offer://?credential_offer=\(offerJson)"
        }
        return "openid-credential-offer://?credential_offer=\(encoded)"
    }
}
